### PR TITLE
Add SampleUri and DetailedTags to SearchResult

### PR DESCRIPTION
**Pull request by Keelhauled**
Some new changes I want to get your opinion on before I work on them further. The testing net5.0 change is just for my own setup but if it's of no concern it could be merged too.

### CODE REVIEW

**Comment by Xwilarg** [comment]
Thanks for your work,
I'm on phone right know so I'll check the whole thing later
However if there is no big additional value, I think it's better to not update to .net5 for compatibility purpose

**Comment by Xwilarg** [comment]
Ah it's for the test project, I need to learn to read

File: BooruSharp.UnitTests/BooruTests.cs [hunk 1]
```diff
@@ -297,8 +297,8 @@ public async Task GetLastPostsWithTagsAsync(Type t, string tag = "hibiki_(kantai
             Assert.NotEqual(results[0].ID, results[1].ID);
             foreach (var elem in results)
             {
-                Assert.Contains(elem.Tags, t => t.Contains(tag));
-                Assert.Contains(elem.Tags, t => t.Contains(tag2));
+                Assert.Contains(elem.Tags, x => x.Contains(tag));
+                Assert.Contains(elem.Tags, x => x.Contains(tag2));
```
Comment (Xwilarg) [inline BooruSharp.UnitTests/BooruTests.cs:301]: What is the reason for this change?

File: BooruSharp.UnitTests/BooruTests.cs [hunk 2]
```diff
@@ -489,6 +489,7 @@ public async Task GetRandomsFailAsync(Type t)
         [InlineData(typeof(E926))]
         //[InlineData(typeof(Furrybooru))]
         [InlineData(typeof(Gelbooru))]
+        [InlineData(typeof(Gelbooru), "'o'ne")]
```
Comment (Xwilarg) [inline BooruSharp.UnitTests/BooruTests.cs:492]: What does the single quotes on the o mean?

File: BooruSharp/Search/Post/SearchResult.cs [hunk 3]
```diff
@@ -73,6 +82,8 @@ namespace BooruSharp.Search.Post
         /// Gets the read-only collection containing all the tags associated with the file.
         /// </summary>
         public ReadOnlyCollection<string> Tags { get; }
+        
+        public ReadOnlyCollection<Tag.SearchResult> DetailedTags { get; }
```
Comment (Xwilarg) [inline BooruSharp/Search/Post/SearchResult.cs:89]: Could you add the XML documentation to explain what this is?

File: BooruSharp/Search/Post/SearchResult.cs [hunk 4]
```diff
@@ -28,15 +30,17 @@ namespace BooruSharp.Search.Post
...
+            IList<Tag.SearchResult> detailedTags, int id, int? size, int height, int width, int? previewHeight, int? previewWidth,
             DateTime? creation, string source, int? score, string md5)
         {
             FileUrl = fileUrl;
             PreviewUrl = previewUrl;
             PostUrl = postUrl;
+            SampleUri = sampleUri;
             Rating = rating;
             Tags = new ReadOnlyCollection<string>(tags);
+            DetailedTags = detailedTags != null ? new ReadOnlyCollection<Tag.SearchResult>(detailedTags) : null;
```
Comment (Xwilarg) [inline BooruSharp/Search/Post/SearchResult.cs:43]: If there is no tag, maybe it would be better to have an empty array instead?

**Review by Xwilarg — CHANGES_REQUESTED** [review]

File: BooruSharp.UnitTests/BooruTests.cs (hunk 2 above)
Comment (Keelhauled) [reply BooruSharp.UnitTests/BooruTests.cs:492]: The quotes are a part of the the tag as you can see [here](https://gelbooru.com/index.php?page=post&s=list&tags=%27o%27ne). Gelbooru returns the tag encoded for some reason so I fixed it.

File: BooruSharp.UnitTests/BooruTests.cs (hunk 1 above)
Comment (Keelhauled) [reply BooruSharp.UnitTests/BooruTests.cs:301]: Naming it t hides the method parameter t. Though now I noticed it's still able to compile it as t so the change is actually pointless.

File: BooruSharp/Search/Post/SearchResult.cs (hunk 3 above)
Comment (Keelhauled) [reply BooruSharp/Search/Post/SearchResult.cs:89]: Absolutely, just stopping to ask if you even like me adding stuff.

File: BooruSharp/Search/Post/SearchResult.cs (hunk 4 above)
Comment (Keelhauled) [reply BooruSharp/Search/Post/SearchResult.cs:43]: Sure

File: BooruSharp.UnitTests/BooruTests.cs (hunk 1 above)
Comment (Xwilarg) [reply BooruSharp.UnitTests/BooruTests.cs:301]: Ah yup I didn't notice on the changes that the parameter was named the same way, nice catch

File: BooruSharp/Search/Post/SearchResult.cs (hunk 3 above)
Comment (Xwilarg) [reply BooruSharp/Search/Post/SearchResult.cs:89]: Ah yup I even forgot to answer that
Yeah, I do appreciate it, thanks for taking time contributing to this project :)

Comment (Keelhauled) [reply BooruSharp/Search/Post/SearchResult.cs:89]: I'll be working on this more then

File: BooruSharp/Booru/ABooru.cs [hunk 5]
```diff
@@ -117,27 +117,27 @@ private protected virtual async Task<IEnumerable> GetTagEnumerableSearchResultAs
...
         /// Gets a value indicating whether tags API uses XML instead of JSON.
         /// </summary>
-        protected bool TagsUseXml => _options.HasFlag(BooruOptions.TagApiXml);
+        public bool TagsUseXml => _options.HasFlag(BooruOptions.TagApiXml);
 
         /// <summary>
         /// Gets a value indicating whether comments API uses XML instead of JSON.
         /// </summary>
-        protected bool CommentsUseXml => _options.HasFlag(BooruOptions.CommentApiXml);
+        public bool CommentsUseXml => _options.HasFlag(BooruOptions.CommentApiXml);
```
Comment (Xwilarg) [inline BooruSharp/Booru/ABooru.cs]: Not sure we want these exposed to the user

File: BooruSharp/Search/Post/SearchResult.cs [hunk 6]
```diff
@@ -15,8 +15,10 @@ namespace BooruSharp.Search.Post
         /// <param name="fileUrl">The URI of the file.</param>
         /// <param name="previewUrl">The URI of the preview image.</param>
         /// <param name="postUrl">The URI of the post.</param>
+        /// <param name="sampleUri">The URI of the sample image.</param>
```
Comment (Xwilarg) [inline BooruSharp/Search/Post/SearchResult.cs]: What is the difference between sample and preview?

**Review by Xwilarg — CHANGES_REQUESTED** [review]

File: BooruSharp/Booru/ABooru.cs (hunk 5 above)
Comment (Keelhauled) [reply BooruSharp/Booru/ABooru.cs]: True, not that one probably. I mainly wanted NoMoreThanTwoTags to avoid the exception with danbooru.

Comment (Xwilarg) [reply BooruSharp/Booru/ABooru.cs]: Where are you using it? If it's only in the library but outside of ABooru you can use internal

File: BooruSharp/Search/Post/SearchResult.cs (hunk 6 above)
Comment (Keelhauled) [reply BooruSharp/Search/Post/SearchResult.cs]: Preview is a thumbnail, sample is a slightly resized version of the full image that's intended to load faster.

File: BooruSharp/Booru/Template/Sankaku.cs [hunk 7]
```diff
@@ -63,6 +81,18 @@ private protected override Search.Post.SearchResult GetPostSearchResult(JToken e
...
 
+        private Search.Tag.TagType GetTagType(int type)
+        {
+            switch(type)
+            {
+                case 1: return Search.Tag.TagType.Artist;
+                case 3: return Search.Tag.TagType.Copyright;
+                case 4: return Search.Tag.TagType.Character;
+                case 8: return Search.Tag.TagType.Metadata;
+                default: return Search.Tag.TagType.Trivia;
```
Comment (Xwilarg) [inline BooruSharp/Booru/Template/Sankaku.cs:92]: I don't know how the API behave but I think it's better to do a case for Trivia and throw an error in default, so if a new tag is created we will know it

**Review by Xwilarg — CHANGES_REQUESTED** [review]

File: BooruSharp/Search/Post/SearchResult.cs (hunk 6 above)
Comment (Xwilarg) [reply BooruSharp/Search/Post/SearchResult.cs]: Do you think that maybe you could explain that in the comment? Would be easier to understand for people that are not aware about it

**Comment by Xwilarg** [comment]
Could you also please add unit tests for both detailled tags and sample URL?

File: BooruSharp/Booru/ABooru.cs (hunk 5 above)
Comment (Keelhauled) [reply BooruSharp/Booru/ABooru.cs]: To clarify, in my own program, I'm using the same code to get posts from all the boorus but any booru with this flag set will cause an exception if I happen to have more than 2 tags. So I need some way to check for it that isn't just `type is Danbooru`.

Comment (Xwilarg) [reply BooruSharp/Booru/ABooru.cs]: Ah makes sence
I think we can keep the more than 2 tags public then, for the others since it's really internal use I think they can go back to protected

File: BooruSharp/Booru/Template/Sankaku.cs (hunk 7 above)
Comment (Keelhauled) [reply BooruSharp/Booru/Template/Sankaku.cs:92]: There are already new tag categories in many boorus but I don't see the value in finding them all when trivia will just do.

File: BooruSharp/Booru/Template/E621.cs [hunk 8]
```diff
@@ -37,33 +38,35 @@ private protected override JToken ParseFirstPostSearchResult(object json)
...
-                "species",
-                "character",
-                "copyright",
-                "artist",
-                "meta",
-            };
+                foreach(var tag in cat.Value.ToObject<string[]>())
+                {
+                    tags.Add(tag);
+                    detailedTags.Add(new Search.Tag.SearchResult(-1, tag, GetTagType(cat.Name), -1));
```
Comment (Xwilarg) [inline BooruSharp/Booru/Template/E621.cs:48]: For detailed tags maybe we could create a simplified version of SearchResult? Since it looks like first and last param are always -1

**Review by Xwilarg — CHANGES_REQUESTED** [review]

File: BooruSharp/Booru/Template/Sankaku.cs (hunk 7 above)
Comment (Xwilarg) [reply BooruSharp/Booru/Template/Sankaku.cs:92]: You only changed tag parsing for Sankaku and E621 or did I miss smth?

Comment (Keelhauled) [reply BooruSharp/Booru/Template/Sankaku.cs:92]: I changed it for all that had tag categories in the response

File: BooruSharp/Booru/Template/E621.cs (hunk 8 above)
Comment (Keelhauled) [reply BooruSharp/Booru/Template/E621.cs:48]: Sankaku has post count and id

**Comment by Keelhauled** [comment]
Maybe you should do the tests later, I can't figure out where and what I should even add

File: BooruSharp/Booru/Template/Sankaku.cs (hunk 7 above)
Comment (Keelhauled) [reply BooruSharp/Booru/Template/Sankaku.cs:92]: All in all I think it's better like this, adding each individual tag category is just a lot of hassle for very little gain. I wanna change gelbooru to do the same too.

File: BooruSharp/Booru/Template/E621.cs (hunk 8 above)
Comment (Xwilarg) [reply BooruSharp/Booru/Template/E621.cs:48]: Ah yeah I didn't notice, my bad

File: BooruSharp/Booru/Template/Sankaku.cs (hunk 7 above)
Comment (Xwilarg) [reply BooruSharp/Booru/Template/Sankaku.cs:92]: Ah I didn't notice E621 change their tag system, Gelbooru still use the "old" one right, what is there to change?

Comment (Xwilarg) [reply BooruSharp/Booru/Template/Sankaku.cs:92]: Uhm... Overall I think adding the others tags could be a neat feature, if it's just that it's a hassle to you, you can let it like that and I'll change this part later myself

Comment (Keelhauled) [reply BooruSharp/Booru/Template/Sankaku.cs:92]: Cool. Gelbooru at the moment will not list any new tag categories that might pop up, instead they are just not even noticed in any way in detailedTags.

**Comment by Xwilarg** [comment]
You can do them here: https://github.com/Xwilarg/BooruSharp/blob/d571c3ff801965d2c73a67b958cb962c63dc125e/BooruSharp.UnitTests/General.cs#L33

File: BooruSharp/Booru/Template/Sankaku.cs (hunk 7 above)
Comment (Keelhauled) [reply BooruSharp/Booru/Template/Sankaku.cs:92]: Oh sorry it was danbooru I was talking about, gelbooru has no tags at all in the response

File: BooruSharp/Search/Post/SearchResult.cs (hunk 4 above)
Comment (Keelhauled) [reply BooruSharp/Search/Post/SearchResult.cs:43]: Maybe detailed tags should be nullable actually, null would mean detailed tags are not even possible and an empty array would mean there are just 0 tags but there might be detailed tags on other images.

Comment (Keelhauled) [reply BooruSharp/Search/Post/SearchResult.cs:43]: Detailed tags are supposed to be a replacement for the string[] of tags so it makes sense to use only one of them and null makes that easy to decide.

Comment (Xwilarg) [reply BooruSharp/Search/Post/SearchResult.cs:43]: Yup I agree

**Comment by Keelhauled** [comment]
I think it's finished

**Review by Xwilarg — APPROVED** [review]

**Comment by Xwilarg** [comment]
Yup, thanks for your contribution

**Comment by Keelhauled** [comment]
And the nuget update?

**Comment by Xwilarg** [comment]
I want to have a quick look at the enum thing before creating the nuget package

**Comment by Xwilarg** [comment]
(Also just noticed that you did the new tag system for detailled tag feature but not for the search tag endpoint)

**Comment by Keelhauled** [comment]
Tag searching is detailed already isn't it? At least GetTagAsync is

**Comment by Xwilarg** [comment]
For example here it uses the old way:

https://github.com/Xwilarg/BooruSharp/blob/master/BooruSharp/Booru/Template/E621.cs#L111-L120

**Comment by Keelhauled** [comment]
I really don't understand what the issue is

**Comment by Xwilarg** [comment]
Ah nevermind it might be me misunderstanding something, need to stop doing many things at once

**Comment by Xwilarg** [comment]
Package will be available here: https://www.nuget.org/packages/BooruSharp/3.3.0
